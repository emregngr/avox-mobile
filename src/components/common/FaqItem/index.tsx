import { MaterialCommunityIcons } from '@expo/vector-icons'
import { useMemo } from 'react'
import { TouchableOpacity, View } from 'react-native'
import Collapsible from 'react-native-collapsible'

import { ThemedText } from '@/components/common/ThemedText'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'
import type { FaqItemType } from '@/types/common/faq'

interface FaqItemProps {
  index: number
  isExpanded: boolean
  item: FaqItemType
  toggleExpanded: (index: number) => void
}

export const FaqItem = ({ index, isExpanded, item, toggleExpanded }: FaqItemProps) => {
  const { selectedTheme } = useThemeStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  const { description, title } = item
  return (
    <View className="mb-2">
      <TouchableOpacity
        activeOpacity={0.7}
        className="p-4 rounded-xl overflow-hidden bg-background-secondary"
        hitSlop={10}
        onPress={() => toggleExpanded(index)}
        testID={`faq-${item?.id}`}
      >
        <View className="flex-row justify-between items-center">
          <View className="flex-1 pr-4">
            <ThemedText color="text-100" type="body1">
              {title}
            </ThemedText>
          </View>
          <MaterialCommunityIcons
            color={colors?.onPrimary50}
            name={isExpanded ? 'chevron-up' : 'chevron-down'}
            size={24}
          />
        </View>
      </TouchableOpacity>

      <Collapsible collapsed={!isExpanded}>
        <View className="mt-2 p-4 rounded-xl overflow-hidden bg-background-tertiary">
          <ThemedText color="text-90" type="body2">
            {description}
          </ThemedText>
        </View>
      </Collapsible>
    </View>
  )
}
