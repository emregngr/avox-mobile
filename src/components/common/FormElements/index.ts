export * from './CheckboxField'
export * from './TextInputField'
