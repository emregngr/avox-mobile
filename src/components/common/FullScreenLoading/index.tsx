import React from 'react'
import { ActivityIndicator, View } from 'react-native'

export const FullScreenLoading = () => (
  <View
    className="flex-1 items-center justify-center bg-background-primary"
    testID="full-screen-loading"
  >
    <ActivityIndicator className="color-text-100" size="large" testID="loading-indicator" />
  </View>
)
