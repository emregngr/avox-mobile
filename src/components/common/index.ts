export * from './FaqItem'
export * from './FormElements'
export * from './FullScreenLoading'
export * from './Header'
export * from './OnboardingItem'
export * from './ProfileItem'
export * from './RenderDetailTabBar'
export * from './RenderTabBar'
export * from './SafeLayout'
export * from './SearchInput'
export * from './ThemedButton'
export * from './ThemedButtonText'
export * from './ThemedGradientButton'
export * from './ThemedText'
export * from './ToastConfig'
