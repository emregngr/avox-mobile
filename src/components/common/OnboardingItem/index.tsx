import { Image } from 'expo-image'
import { View } from 'react-native'

import { ThemedText } from '@/components/common/ThemedText'
import type { OnBoardingType } from '@/types/common/onBoarding'

interface OnboardingItemProps {
  item: OnBoardingType
}

const STATIC_STYLES = {
  image: {
    height: '45%' as const,
    marginVertical: 20,
    width: '100%' as const,
  },
}

export const OnboardingItem = ({ item }: OnboardingItemProps) => {
  const { image, text, title } = item
  return (
    <View className="w-screen h-screen bg-background-primary">
      <Image
        cachePolicy="memory-disk"
        contentFit="contain"
        source={image}
        style={STATIC_STYLES.image}
        transition={0}
      />
      <View className="mx-4 z-10">
        <ThemedText
          className="mb-4"
          color="text-100"
          lineBreakMode="tail"
          numberOfLines={2}
          type="h1"
        >
          {title}
        </ThemedText>
        <ThemedText
          color="text-100" lineBreakMode="tail" numberOfLines={4}
          type="h3"
        >
          {text}
        </ThemedText>
      </View>
    </View>
  )
}
