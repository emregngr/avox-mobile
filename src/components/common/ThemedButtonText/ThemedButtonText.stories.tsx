import { MaterialCommunityIcons } from '@expo/vector-icons'
import type { Meta, StoryObj } from '@storybook/react-native'
import React from 'react'
import { Alert, View } from 'react-native'

import { ThemedButtonText } from '@/components/common/ThemedButtonText'
import { ThemedText } from '@/components/common/ThemedText'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'

const meta: Meta<typeof ThemedButtonText> = {
  title: 'Components/ThemedButtonText',
  component: ThemedButtonText,
  parameters: {
    docs: {
      description: {
        component:
          'A themed button component that displays text with optional icon and customizable typography.',
      },
    },
  },
  argTypes: {
    label: {
      control: 'text',
      description: 'Button text label',
    },
    type: {
      control: 'select',
      options: [
        'bigTitle',
        'title',
        'h1',
        'h2',
        'h3',
        'h4',
        'body1',
        'body2',
        'body3',
        'body4',
        'button1',
        'button2',
        'tabBar',
      ],
      description: 'Typography type',
    },
    textColor: {
      control: 'text',
      description: 'Text color class',
    },
    disabled: {
      control: 'boolean',
      description: 'Disable button interaction',
    },
    hitSlop: {
      control: 'number',
      description: 'Touch area extension',
    },
    containerStyle: {
      control: 'text',
      description: 'Additional container styles',
    },
    textStyle: {
      control: 'text',
      description: 'Additional text styles',
    },
    onPress: {
      action: 'pressed',
      description: 'Press handler function',
    },
  },
  decorators: [
    Story => (
      <View className="flex-1 p-4 bg-background-primary">
        <Story />
      </View>
    ),
  ],
}

export default meta

type Story = StoryObj<typeof ThemedButtonText>

export const Default: Story = {
  args: {
    label: 'Default Button',
    textColor: 'primary-100',
    onPress: () => Alert.alert('Button Pressed', 'Default button was pressed!'),
  },
}

export const TypographyVariations: Story = {
  render: () => (
    <View className="gap-4">
      <ThemedButtonText
        label="Big Title Button"
        type="bigTitle"
        textColor="text-70"
        onPress={() => Alert.alert('Big Title')}
      />
      <ThemedButtonText
        label="Title Button"
        type="title"
        textColor="text-70"
        onPress={() => Alert.alert('Title')}
      />
      <ThemedButtonText
        label="H1 Button"
        type="h1"
        textColor="text-70"
        onPress={() => Alert.alert('H1')}
      />
      <ThemedButtonText
        label="Body1 Button"
        type="body1"
        textColor="text-70"
        onPress={() => Alert.alert('Body1')}
      />
      <ThemedButtonText
        label="Button1 Style"
        type="button1"
        textColor="primary-100"
        onPress={() => Alert.alert('Button1')}
      />
      <ThemedButtonText
        label="Button2 Style"
        type="button2"
        textColor="primary-100"
        onPress={() => Alert.alert('Button2')}
      />
    </View>
  ),
}

export const ColorVariations: Story = {
  render: () => (
    <View className="gap-4">
      <ThemedButtonText
        label="Primary Color"
        textColor="primary-100"
        onPress={() => Alert.alert('Primary')}
      />
      <ThemedButtonText
        label="Secondary Color"
        textColor="secondary-100"
        onPress={() => Alert.alert('Secondary')}
      />
      <ThemedButtonText
        label="Success Color"
        textColor="success"
        onPress={() => Alert.alert('Success')}
      />
      <ThemedButtonText
        label="Warning Color"
        textColor="warning"
        onPress={() => Alert.alert('Warning')}
      />
      <ThemedButtonText
        label="Danger Color"
        textColor="error"
        onPress={() => Alert.alert('Danger')}
      />
    </View>
  ),
}

export const WithIcons: Story = {
  render: () => {
    const { selectedTheme } = useThemeStore()
    const colors = themeColors?.[selectedTheme]

    return (
      <View className="gap-4">
        <ThemedButtonText
          label="Download"
          textColor="text-100"
          icon={<MaterialCommunityIcons name="download" size={20} color={colors?.primary100} />}
          onPress={() => Alert.alert('Download')}
        />
        <ThemedButtonText
          label="Share"
          textColor="text-100"
          icon={<MaterialCommunityIcons name="share" size={20} color={colors?.success} />}
          onPress={() => Alert.alert('Share')}
        />
        <ThemedButtonText
          label="Delete"
          textColor="text-100"
          icon={<MaterialCommunityIcons name="delete" size={20} color={colors?.error} />}
          onPress={() => Alert.alert('Delete')}
        />
        <ThemedButtonText
          label="File"
          textColor="text-100"
          icon={<MaterialCommunityIcons name="file" size={20} color={colors?.secondary100} />}
          onPress={() => Alert.alert('File')}
        />
      </View>
    )
  },
}

export const Disabled: Story = {
  render: () => {
    const { selectedTheme } = useThemeStore()
    const colors = themeColors?.[selectedTheme]

    return (
      <View className="gap-4">
        <ThemedButtonText
          label="Enabled Button"
          textColor="text-100"
          onPress={() => Alert.alert('Enabled')}
        />
        <ThemedButtonText
          label="Disabled Button"
          textColor="text-50"
          disabled={true}
          onPress={() => Alert.alert('This should not show')}
        />
        <ThemedButtonText
          label="Disabled with Icon"
          textColor="text-50"
          disabled={true}
          icon={
            <MaterialCommunityIcons name="lock" size={20} color={colors?.background?.quaternary} />
          }
          onPress={() => Alert.alert('This should not show')}
        />
      </View>
    )
  },
}

export const CustomStyling: Story = {
  render: () => (
    <View className="gap-4">
      <ThemedButtonText
        label="Custom Container"
        textColor="text-100"
        containerStyle="bg-primary-100 px-6 py-3 rounded-lg"
        onPress={() => Alert.alert('Custom Container')}
      />
      <ThemedButtonText
        label="Custom Text Style"
        textColor="primary-100"
        textStyle="font-bold underline"
        onPress={() => Alert.alert('Custom Text')}
      />
      <ThemedButtonText
        label="Centered Button"
        textColor="text-100"
        containerStyle="justify-center bg-primary-100 py-4 rounded-full"
        onPress={() => Alert.alert('Centered')}
      />
    </View>
  ),
}

export const Interactive: Story = {
  args: {
    label: 'Interactive Button',
    type: 'button1',
    textColor: 'text-100',
    disabled: false,
    hitSlop: 20,
    containerStyle: 'bg-primary-100 px-4 py-2 rounded-md',
    textStyle: 'font-medium',
    onPress: () => Alert.alert('Interactive', 'You can modify props in the controls!'),
  },
}

export const HitSlopDemo: Story = {
  render: () => (
    <View className="gap-8 items-center">
      <View>
        <ThemedText type="body2" color="text-70">
          Small hit area (hitSlop: 5)
        </ThemedText>
        <ThemedButtonText
          label="Hard to tap"
          textColor="error"
          hitSlop={5}
          onPress={() => Alert.alert('Small hit area')}
        />
      </View>

      <View>
        <ThemedText type="body2" color="text-70">
          Large hit area (hitSlop: 40)
        </ThemedText>
        <ThemedButtonText
          label="Easy to tap"
          textColor="success"
          hitSlop={40}
          onPress={() => Alert.alert('Large hit area')}
        />
      </View>
    </View>
  ),
}
