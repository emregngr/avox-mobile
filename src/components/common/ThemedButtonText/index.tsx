import type { ReactNode } from 'react'
import React from 'react'
import { Pressable, View } from 'react-native'

import { ThemedText } from '@/components/common/ThemedText'
import { cn } from '@/utils/common/cn'

type TypographyType =
  | 'bigTitle'
  | 'title'
  | 'h1'
  | 'h2'
  | 'h3'
  | 'h4'
  | 'body1'
  | 'body2'
  | 'body3'
  | 'body4'
  | 'button1'
  | 'button2'
  | 'tabBar'

type ThemedButtonTextProps = {
  [key: string]: any
  containerStyle?: string
  icon?: ReactNode
  label: string
  onPress?: () => void
  textColor: string
  textStyle?: string
  type?: TypographyType
}

export const ThemedButtonText = ({
  containerStyle,
  icon,
  label,
  textColor = 'text-100',
  textStyle,
  type = 'button2',
  ...rest
}: ThemedButtonTextProps) => (
  <Pressable className={cn('flex-row', containerStyle)} hitSlop={30} {...rest}>
    <ThemedText className={textStyle as string} color={textColor} type={type}>
      {label}
    </ThemedText>
    {icon ? <View className="ml-2.5">{icon}</View> : null}
  </Pressable>
)
