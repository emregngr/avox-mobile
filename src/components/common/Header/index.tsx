import { GlassView } from 'expo-glass-effect'
import * as Haptics from 'expo-haptics'
import { type ReactNode, useMemo } from 'react'
import type { ViewStyle } from 'react-native'
import { TouchableOpacity, View } from 'react-native'

import Back from '@/assets/icons/back.svg'
import { ThemedButtonText } from '@/components/common/ThemedButtonText'
import { ThemedText } from '@/components/common/ThemedText'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'
import { cn } from '@/utils/common/cn'

type HeaderProps = {
  backIcon?: boolean
  backIconOnPress?: () => void
  containerClassName?: string
  hapticFeedback?: boolean
  isFavorite?: boolean
  rightButtonLabel?: string
  rightButtonOnPress?: () => void
  rightIcon?: ReactNode
  rightIconOnPress?: () => void
  shareIcon?: ReactNode
  shareIconOnPress?: () => void
  style?: ViewStyle
  testID?: string
  title?: string | string[]
  titleClassName?: string
}

export const Header = ({
  backIcon = true,
  backIconOnPress,
  containerClassName,
  hapticFeedback = false,
  isFavorite = false,
  rightButtonLabel,
  rightButtonOnPress,
  rightIcon,
  rightIconOnPress,
  shareIcon,
  shareIconOnPress,
  style,
  testID,
  title,
  titleClassName,
}: HeaderProps) => {
  const { selectedTheme } = useThemeStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  const STATIC_STYLES = {
    backIcon: {
      left: 16,
      position: 'absolute' as const,
      borderRadius: 20,
      padding: 8,
    },
    shareIcon: {
      right: 64,
      position: 'absolute' as const,
      borderRadius: 20,
      padding: 8,
    },
    rightIcon: {
      right: 16,
      position: 'absolute' as const,
      borderRadius: 20,
      padding: 8,
    },
  }

  const handleRightIconOnPress = () => {
    if (hapticFeedback) {
      if (isFavorite) {
        Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Light)
      } else {
        Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Medium)
      }
    }

    rightIconOnPress?.()
  }

  return (
    <View
      className={cn('h-11 justify-center items-center', containerClassName)}
      style={style}
      testID={testID || 'header'}
    >
      {title ? (
        <ThemedText
          className={titleClassName ?? ''}
          color="text-100"
          ellipsizeMode="tail"
          numberOfLines={2}
          type="h3"
          center
        >
          {title}
        </ThemedText>
      ) : null}

      {backIcon ? (
        <GlassView
          style={STATIC_STYLES.backIcon}
          tintColor={colors?.background?.glass}
          glassEffectStyle="clear"
          isInteractive
        >
          <TouchableOpacity
            activeOpacity={0.7}
            hitSlop={20}
            onPress={backIconOnPress}
            testID="header-back-icon"
          >
            <Back color={colors?.onPrimary100} height={24} width={24} />
          </TouchableOpacity>
        </GlassView>
      ) : null}

      {shareIcon ? (
        <GlassView
          style={STATIC_STYLES.shareIcon}
          tintColor={colors?.background?.glass}
          glassEffectStyle="clear"
          isInteractive
        >
          <TouchableOpacity
            activeOpacity={0.7}
            hitSlop={10}
            onPress={shareIconOnPress}
            testID="header-share-icon"
          >
            {shareIcon}
          </TouchableOpacity>
        </GlassView>
      ) : null}

      {rightButtonLabel ? (
        <ThemedButtonText
          containerStyle="absolute right-4"
          label={rightButtonLabel}
          onPress={rightButtonOnPress as () => void}
          textColor="text-100"
          type="h4"
        />
      ) : rightIcon ? (
        <GlassView
          style={STATIC_STYLES.rightIcon}
          tintColor={colors?.background?.glass}
          glassEffectStyle="clear"
          isInteractive
        >
          <TouchableOpacity
            activeOpacity={0.7}
            hitSlop={10}
            onPress={handleRightIconOnPress}
            testID="header-right-icon"
          >
            {rightIcon}
          </TouchableOpacity>
        </GlassView>
      ) : null}
    </View>
  )
}
