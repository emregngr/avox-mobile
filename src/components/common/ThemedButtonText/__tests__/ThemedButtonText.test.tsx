import { fireEvent, render } from '@testing-library/react-native'
import React from 'react'
import { View } from 'react-native'

import { ThemedButtonText } from '@/components/common/ThemedButtonText'

jest.mock('@/components/common/ThemedText', () => {
  const { Text } = require('react-native')

  return {
    ThemedText: ({ children, ...props }: { children: string }) => (
      <Text testID="themed-text" {...props}>
        {children}
      </Text>
    ),
  }
})

jest.mock('@/utils/common/cn', () => ({
  cn: (...classes: (string | undefined)[]) => classes.filter(Boolean).join(' '),
}))

const mockedRequiredProps = {
  label: 'Test Button',
  textColor: 'text-100',
}

describe('ThemedButtonText Component', () => {
  it('renders correctly with required props', () => {
    const { getByText, getByTestId } = render(<ThemedButtonText {...mockedRequiredProps} />)

    expect(getByText('Test Button')).toBeTruthy()
    expect(getByTestId('themed-text')).toBeTruthy()
  })

  it('renders with default values', () => {
    const { getByTestId } = render(<ThemedButtonText {...mockedRequiredProps} />)

    const themedText = getByTestId('themed-text')
    expect(themedText).toBeTruthy()
  })

  describe('Props Handling', () => {
    it('applies custom containerStyle', () => {
      const { getByTestId } = render(
        <ThemedButtonText
          {...mockedRequiredProps}
          containerStyle="custom-container-style"
          testID="pressable-button"
        />,
      )

      const pressable = getByTestId('pressable-button')
      expect(pressable).toBeTruthy()
    })

    it('applies custom textStyle', () => {
      const { getByTestId } = render(
        <ThemedButtonText {...mockedRequiredProps} textStyle="custom-text-style" />,
      )

      const themedText = getByTestId('themed-text')
      expect(themedText.props.className).toBe('custom-text-style')
    })

    it('uses custom typography type', () => {
      const { getByTestId } = render(<ThemedButtonText {...mockedRequiredProps} type="h1" />)

      const themedText = getByTestId('themed-text')
      expect(themedText.props.type).toBe('h1')
    })

    it('uses custom textColor', () => {
      const { getByTestId } = render(
        <ThemedButtonText {...mockedRequiredProps} textColor="text-primary" />,
      )

      const themedText = getByTestId('themed-text')
      expect(themedText.props.color).toBe('text-primary')
    })

    it('applies custom hitSlop', () => {
      const { getByTestId } = render(
        <ThemedButtonText {...mockedRequiredProps} hitSlop={30} testID="pressable-button" />,
      )

      const pressable = getByTestId('pressable-button')
      expect(pressable.props.hitSlop).toBe(30)
    })
  })

  describe('Icon Rendering', () => {
    it('renders without icon by default', () => {
      const { queryByTestId } = render(<ThemedButtonText {...mockedRequiredProps} />)

      expect(queryByTestId('test-icon')).toBeFalsy()
    })

    it('renders with icon when provided', () => {
      const icon = <View testID="test-icon" />
      const { getByTestId } = render(<ThemedButtonText {...mockedRequiredProps} icon={icon} />)

      expect(getByTestId('test-icon')).toBeTruthy()
    })
  })

  describe('Press Handling', () => {
    it('calls onPress when pressed', () => {
      const mockedOnPressMock = jest.fn()
      const { getByTestId } = render(
        <ThemedButtonText
          {...mockedRequiredProps}
          onPress={mockedOnPressMock}
          testID="pressable-button"
        />,
      )

      fireEvent.press(getByTestId('pressable-button'))
      expect(mockedOnPressMock).toHaveBeenCalledTimes(1)
    })

    it('does not crash when onPress is not provided', () => {
      const { getByTestId } = render(
        <ThemedButtonText {...mockedRequiredProps} testID="pressable-button" />,
      )

      expect(() => {
        fireEvent.press(getByTestId('pressable-button'))
      }).not.toThrow()
    })
  })

  describe('Rest Props', () => {
    it('passes through additional Pressable props', () => {
      const { getByTestId } = render(
        <ThemedButtonText
          {...mockedRequiredProps}
          accessibilityLabel="Custom accessibility label"
          accessibilityRole="button"
          testID="custom-test-id"
        />,
      )

      const pressable = getByTestId('custom-test-id')
      expect(pressable.props.accessibilityLabel).toBe('Custom accessibility label')
      expect(pressable.props.accessibilityRole).toBe('button')
    })
  })
})

describe('ThemedButtonText Component Snapshot', () => {
  it('should render the ThemedButtonText Component successfully', () => {
    const { toJSON } = render(<ThemedButtonText {...mockedRequiredProps} />)

    expect(toJSON()).toMatchSnapshot()
  })
})
