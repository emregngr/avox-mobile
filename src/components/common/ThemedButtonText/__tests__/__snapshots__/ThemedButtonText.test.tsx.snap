// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ThemedButtonText Component Snapshot should render the ThemedButtonText Component successfully 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  className="flex-row items-center"
  collapsable={false}
  focusable={true}
  hitSlop={20}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <Text
    color="text-100"
    testID="themed-text"
    type="button2"
  >
    Test Button
  </Text>
</View>
`;
